File: audiobook/src/main/java/de/ph1b/audiobook/injection/ApplicationComponent.java
package de.ph1b.audiobook.injection;

import android.content.Context;

import javax.inject.Singleton;

import dagger.Component;
import de.ph1b.audiobook.features.BaseActivity;
import de.ph1b.audiobook.features.BookActivity;
import de.ph1b.audiobook.features.book_overview.BookShelfAdapter;
import de.ph1b.audiobook.features.book_overview.BookShelfController;
import de.ph1b.audiobook.features.book_overview.BookShelfPresenter;
import de.ph1b.audiobook.features.book_overview.EditBookBottomSheet;
import de.ph1b.audiobook.features.book_overview.EditBookTitleDialogFragment;
import de.ph1b.audiobook.features.book_playing.BookPlayController;
import de.ph1b.audiobook.features.book_playing.JumpToPositionDialogFragment;
import de.ph1b.audiobook.features.book_playing.SeekDialogFragment;
import de.ph1b.audiobook.features.book_playing.SleepTimerDialogFragment;
import de.ph1b.audiobook.features.bookmarks.BookmarkDialogFragment;
import de.ph1b.audiobook.features.folder_chooser.FolderChooserActivity;
import de.ph1b.audiobook.features.folder_chooser.FolderChooserPresenter;
import de.ph1b.audiobook.features.folder_overview.FolderOverviewPresenter;
import de.ph1b.audiobook.features.imagepicker.ImagePickerController;
import de.ph1b.audiobook.features.settings.SettingsController;
import de.ph1b.audiobook.features.settings.dialogs.AutoRewindDialogFragment;
import de.ph1b.audiobook.features.settings.dialogs.PlaybackSpeedDialogFragment;
import de.ph1b.audiobook.features.settings.dialogs.ThemePickerDialogFragment;
import de.ph1b.audiobook.features.widget.WidgetUpdateService;
import de.ph1b.audiobook.playback.PlaybackService;
import de.ph1b.audiobook.playback.utils.ChangeNotifier;
import de.ph1b.audiobook.uitools.CoverReplacement;

/**
 * Base component that is the entry point for injection.
 *
 * @author Paul Woitaschek
 */
@SuppressWarnings("WeakerAccess")
@Singleton @Component(modules = {BaseModule.class, AndroidModule.class, PresenterModule.class, PrefsModule.class})
public interface ApplicationComponent {

   BookShelfPresenter getBookShelfPresenter();
   Context getContext();

   void inject(App target);
   void inject(AutoRewindDialogFragment target);
   void inject(BaseActivity target);
   void inject(PlaybackService target);
   void inject(BookActivity target);
   void inject(BookShelfAdapter target);
   void inject(BookmarkDialogFragment target);
   void inject(BookPlayController target);
   void inject(BookShelfController target);
   void inject(ChangeNotifier target);
   void inject(CoverReplacement target);
   void inject(SettingsController target);
   void inject(EditBookTitleDialogFragment target);
   void inject(EditBookBottomSheet target);
   void inject(FolderChooserActivity target);
   void inject(FolderChooserPresenter target);
   void inject(FolderOverviewPresenter target);
   void inject(ImagePickerController target);
   void inject(JumpToPositionDialogFragment target);
   void inject(PlaybackSpeedDialogFragment target);
   void inject(SeekDialogFragment target);
   void inject(SleepTimerDialogFragment target);
   void inject(ThemePickerDialogFragment target);
   void inject(WidgetUpdateService target);
}

File: audiobook/src/main/java/de/ph1b/audiobook/injection/ApplicationComponent.kt
package de.ph1b.audiobook.injection

import android.content.Context
import dagger.Component
import de.ph1b.audiobook.features.BaseActivity
import de.ph1b.audiobook.features.MainActivity
import de.ph1b.audiobook.features.bookOverview.*
import de.ph1b.audiobook.features.bookPlaying.*
import de.ph1b.audiobook.features.bookmarks.BookmarkDialogFragment
import de.ph1b.audiobook.features.folderChooser.FolderChooserActivity
import de.ph1b.audiobook.features.folderChooser.FolderChooserPresenter
import de.ph1b.audiobook.features.folderOverview.FolderOverviewPresenter
import de.ph1b.audiobook.features.imagepicker.ImagePickerController
import de.ph1b.audiobook.features.settings.SettingsController
import de.ph1b.audiobook.features.settings.dialogs.AutoRewindDialogFragment
import de.ph1b.audiobook.features.settings.dialogs.PlaybackSpeedDialogFragment
import de.ph1b.audiobook.features.settings.dialogs.ThemePickerDialogFragment
import de.ph1b.audiobook.features.widget.WidgetUpdateService
import de.ph1b.audiobook.persistence.PrefsManager
import de.ph1b.audiobook.playback.MediaPlayer
import de.ph1b.audiobook.playback.PlayStateManager
import de.ph1b.audiobook.playback.PlaybackService
import de.ph1b.audiobook.playback.utils.ChangeNotifier
import javax.inject.Singleton

/**
 * Base component that is the entry point for injection.
 *
 * @author Paul Woitaschek
 */
@Singleton
@Component(modules = arrayOf(AndroidModule::class, PrefsModule::class))
interface ApplicationComponent {

  val bookShelfPresenter: BookShelfPresenter
  val context: Context
  val prefsManager: PrefsManager
  val player: MediaPlayer
  val playStateManager: PlayStateManager

  fun inject(target: App)
  fun inject(target: AutoRewindDialogFragment)
  fun inject(target: BaseActivity)
  fun inject(target: BookmarkDialogFragment)
  fun inject(target: BookPlayController)
  fun inject(target: BookPlayPresenter)
  fun inject(target: BookShelfAdapter)
  fun inject(target: BookShelfController)
  fun inject(target: ChangeNotifier)
  fun inject(target: EditBookBottomSheet)
  fun inject(target: EditBookTitleDialogFragment)
  fun inject(target: EditCoverDialogFragment)
  fun inject(target: FolderChooserActivity)
  fun inject(target: FolderChooserPresenter)
  fun inject(target: FolderOverviewPresenter)
  fun inject(target: ImagePickerController)
  fun inject(target: JumpToPositionDialogFragment)
  fun inject(target: MainActivity)
  fun inject(target: PlaybackService)
  fun inject(target: PlaybackSpeedDialogFragment)
  fun inject(target: SeekDialogFragment)
  fun inject(target: SettingsController)
  fun inject(target: SleepTimerDialogFragment)
  fun inject(target: ThemePickerDialogFragment)
  fun inject(target: WidgetUpdateService)
}